### ISSUE TITLE
service.dead now only operates if the service file exists

### ISSUE THREAD

**Issue by jdelic** [saltstack/salt#37511]
# Issue

Previous to 2016.3.4 it was possible to have a `service.dead` state for a service that didn't exist on the system in the first place. 2016.3.4 now reports "The named service XYZ is not available". This makes it impossible to use Salt to ensure that a service is either disabled *or* not installed. It can now only be disabled.

In my repository I have two service states for Hashicorp consul. One for the "consul agent" and one for the "consul server". A specific node should only ever have one of these installed, so in my Salt configuration the "consul agent" configuration contains a `service.dead` clause for the "consul server" systemd service and the "consul server" configuration contains a `service.dead` clause for the "consul agent" systemd service.

My expected behavior was that if the node was assigned *one* of the states and the other state was previously installed, it would be shutdown and disabled. If *both* states got accidentally assigned, Salt would detect a conflict and report an error.

The above pattern is now broken. There seems to be no way for `service.dead` to also succeed when a service is not installed in the first place.

## Potential solution
Add a flag `must_exist` to `service.dead`, `service.enable` and `service.disable` allowing Salt to *enable*, *disable* and *kill and disable* services only if they exist.

**Comment by Ch3LL** [comment]
@jdelic i definitely like your approach to add an argument. Looks like you are working on a fix. Please feel free to submit that as a PR. Thanks

**Comment by donfaruco** [comment]
I'm not sure if this is completely related but similar.  Hoping you can help.  As of this weekend I started having a state fail on my deployment process.  This appears to be related to the new release of salt-minion 2016.3.4.  The output I get s below.  I have an init script called /etc/init.d/forever.  This state has always just started the service for me.  Why the servie does start and works ok.  The state shows as failure with below output making my deployment pipeline fail due to the error.  Is there something I should be doing different with this new version?

SLS file:
forever:
  service.running:
    - enable: True

Output:
local:
----------
          ID: forever-service
    Function: service.running
        Name: forever
      Result: False
     Comment: Service forever is already enabled, and is dead
     Started: 14:16:51.269911
    Duration: 784.467 ms
     Changes:

Summary for local
------------
Succeeded: 0
Failed:    1

**Comment by jdelic** [comment]
> Please feel free to submit that as a PR. Thanks

Wish that I were, but I know far too less about Salt internals to do so. Github just associated work-arounds that I committed to my Salt configurations with this ticket, because I had this ticket in the commit comments. Sorry 😞

**Comment by donfaruco** [comment]
Ok I found the problem why the salt state is returning the following comment and failing the state.

**Comment: Service forever is already enabled, and is dead**

According to the salt service module note at https://docs.saltstack.com/en/latest/ref/states/all/salt.states.service.html#starting-or-restarting-of-services-and-daemons

The current status of a service is determined by the return code of the init/rc script **status** command.  My forever init script is a custom script and does not have a status option.  Therefore while it start the service it never report the correct status to return a success code therefore failing the state.  Wonder if there is a way to tell it to handle that more gracefully.

Thanks all.

**Comment by terminalmage** [comment]
@jdelic I wish you had provided more information, I simply cannot reproduce this in 2016.3.3:

```
# salt-call --version
salt-call 2016.3.3 (Boron)
# salt-call --local state.single service.dead asdf
11:37:24,620 [py.warnings][2216][WARNING ][23611] /home/erik/git/salt/main/salt/grains/core.py:1493: DeprecationWarning: The "osmajorrelease" will be a type of an integer.

11:37:24,851 [salt.state][ 245][ERROR   ][23611] Failed to get information on unit 'asdf'
local:
----------
          ID: asdf
    Function: service.dead
      Result: False
     Comment: Failed to get information on unit 'asdf'
     Started: 11:37:24.842197
    Duration: 9.407 ms
     Changes:

Summary for local
------------
Succeeded: 0
Failed:    1
------------
Total states run:     1
Total run time:   9.407 ms
```

It seems like you completely bypassed providing any of the information requested in the issue template. Could you please provide this to help us investigate?

**Comment by terminalmage** [comment]
I tried every single release going back to the 2015.8 release cycle and it seems like this behavior changed between 2015.8.7 and 2015.8.8.

**Comment by jdelic** [comment]
@terminalmage I apologize, the command sequence you wrote is precisely what I arrived at to "reproduce the problem", so yes, I didn't use the issue template, I didn't think I missed any important information in my writeup. Now, while working on #37545 I realized the following:

  * For about a year now my configuration worked like I described in this ticket using `service.dead` to make sure that either "consul agent" or the other service "consul agent -server" was running, never both.
  * Perhaps this should have **never** worked (even though I remember it working) because `service.dead` would have always failed when one of the services wasn't available to systemd.
  * However, due to the reactor not syncing grains with my salt-master before the second highstate (see #37545), *all* my nodes were configured with the "consul agent" service that was then later killed by `service.dead`, so my configuration worked even though it shouldn't have. Some of the nodes should have been provisioned with "consul agent -server" from the start.

Long story short: This might not be a bug. I will file a RFE for adding a "only_if_exists" flag to `salt.states.service.*` because I think the pattern would be generally useful, but this ticket is likely symptom, not a bug.

Thank you for investigating, though.

**Comment by jdelic** [comment]
> I tried every single release going back to the 2015.8 release cycle and it seems like this behavior changed between 2015.8.7 and 2015.8.8.

ha! that would explain that I remember this pattern working... but my current problems don't stem from this behavior change. :rocket:

**Comment by terminalmage** [comment]
No, this definitely seems like behavior that changed at some point. Since it doesn't appear to have been intentional (I see no documentation of the change), I think this should be considered a bug.

It looks like we also have a second wrinkle, and that is the the fact that for nonexistant units (in systemd), systemd.py raises an exception, and when that exception is caught we treat the state as failed.

**Comment by terminalmage** [comment]
OK, upon further investigation, it seems that this exception is only raised in earlier Salt releases when the minion is running systemd 231, since the output of ``systemctl status`` changed in that version. My laptop is running Arch and systemd 231, so that issue with the error being raised was due to my systemd version. When repeating the test on CentOS 7, the state succeeds on 2016.3.3 and fails on 2016.3.4.

It looks like the breakage occurred [here](https://github.com/saltstack/salt/commit/c4f899b#diff-51f5a9819bcf2919b421cb7fccd2c42bR431). I'm working on a fix, as well as more unit/integration tests to catch cases like this.

**Comment by danlsgiga** [comment]
Great catch @terminalmage. Thanks for troubleshooting this! Even though I love Arch (I use it in my laptop), I don't think anyone uses it in the server side which might cause caveats like this one.

**Comment by terminalmage** [comment]
Yeah, though RedHat is backporting these systemd changes for RHEL 7.3, even though their systemd version will remain at 219.

**Comment by terminalmage** [comment]
Fix submitted in https://github.com/saltstack/salt/pull/37562 along with a new unit test.
